app/Rules/UsernameFormatRule.php

```diff
@@ -5,13 +5,56 @@
 use App\Actions\GenerateUsername;
 use Closure;
 use Illuminate\Contracts\Validation\ValidationRule;
+use Illuminate\Contracts\Validation\ValidatorAwareRule;
+use Illuminate\Support\Facades\Validator as FacadeValidator;
+use Illuminate\Validation\Validator;
 
-class UsernameFormatRule implements ValidationRule
+class UsernameFormatRule implements ValidationRule, ValidatorAwareRule
 {
+    protected Validator $validator;
+
+    protected array $rules = [
+        'required',
+        'string',
+        'min:2',
+        'max:50',
+        'unique:users,username',
+    ];
+
     public function validate(string $attribute, mixed $value, Closure $fail): void
     {
-        if (! is_string($value) || $value !== (new GenerateUsername())($value)) {
-            $fail('The :attribute must be valid');
+        $validator = FacadeValidator::make(
+            [$attribute => $value],
+            [$attribute => $this->rules],
+        )->after(function () use ($value, $fail) {
+            if (! is_string($value)) {
+                return;
+            }
+
+            if ($value !== (new GenerateUsername())($value)) {
+                $fail(
+                    __(
+                        'The :attribute must be valid. Only lowercase ASCII characters are allowed. Hyphens can be used. Whitespace, underscores, and multiple hyphens are not permitted. The username cannot start or end with a hyphen.'
```

Comment:
Could we rewrite it to something a bit shorter?

> The :attribute must be valid. Only lowercase characters and non-repeating hyphens (`-`) are allowed.

Reply:
Ideally, we'd provide a separate message if the username starts or ends with a hyphen:

> The :attribute must be valid. You cannot use a hyphen (`-`) at the start or end of the :attribute.

Reply:
Ok I'll look in to it, thanks.

Reply:
Ok I shorten the error message, for the second part, the one about not allowing usernames to start with `_` or `-` I used Laravel's [doesnt_start_with:](https://laravel.com/docs/10.x/validation#rule-doesnt-start-with) and [doesnt_end_with](https://laravel.com/docs/10.x/validation#rule-doesnt-end-with)